src/client/ws/tests.rs

```diff
@@ -1 +1,125 @@
+	assert_error_response(response, jsonrpc::ErrorCode::ParseError, PARSE_ERROR.into());
+}
+
+#[tokio::test]
+async fn invalid_request_works() {
+	let server =
+		WebSocketTestServer::with_hardcoded_response("127.0.0.1:0".parse().unwrap(), invalid_request(Id::Num(0_u64)))
+			.await;
+	let uri = to_ws_uri_string(server.local_addr());
+	let client = WsClient::new(&uri, WsConfig::default()).await.unwrap();
+	let response: Result<jsonrpc::JsonValue, Error> = client.request("say_hello", jsonrpc::Params::None).await;
+	assert_error_response(response, jsonrpc::ErrorCode::InvalidRequest, INVALID_REQUEST.into());
+}
+
+#[tokio::test]
+async fn invalid_params_works() {
+	let server =
+		WebSocketTestServer::with_hardcoded_response("127.0.0.1:0".parse().unwrap(), invalid_params(Id::Num(0_u64)))
+			.await;
+	let uri = to_ws_uri_string(server.local_addr());
+	let client = WsClient::new(&uri, WsConfig::default()).await.unwrap();
+	let response: Result<jsonrpc::JsonValue, Error> = client.request("say_hello", jsonrpc::Params::None).await;
+	assert_error_response(response, jsonrpc::ErrorCode::InvalidParams, INVALID_PARAMS.into());
+}
+
+#[tokio::test]
+async fn internal_error_works() {
+	let server =
+		WebSocketTestServer::with_hardcoded_response("127.0.0.1:0".parse().unwrap(), internal_error(Id::Num(0_u64)))
+			.await;
+	let uri = to_ws_uri_string(server.local_addr());
+	let client = WsClient::new(&uri, WsConfig::default()).await.unwrap();
+	let response: Result<jsonrpc::JsonValue, Error> = client.request("say_hello", jsonrpc::Params::None).await;
+	assert_error_response(response, jsonrpc::ErrorCode::InternalError, INTERNAL_ERROR.into());
+}
+
+#[tokio::test]
+async fn subscription_works() {
+	let server = WebSocketTestServer::with_hardcoded_subscription(
+		"127.0.0.1:0".parse().unwrap(),
+		server_subscription_id_response(Id::Num(0)),
+		server_subscription_response(jsonrpc::JsonValue::String("hello my friend".to_owned())),
+	)
+	.await;
+	let uri = to_ws_uri_string(server.local_addr());
+	let client = WsClient::new(&uri, WsConfig::default()).await.unwrap();
+	{
+		let mut sub: WsSubscription<String> =
+			client.subscribe("subscribe_hello", jsonrpc::Params::None, "unsubscribe_hello").await.unwrap();
+		let response: String = sub.next().await.unwrap().into();
+		assert_eq!("hello my friend".to_owned(), response);
+	}
+}
+
+#[tokio::test]
+async fn response_with_wrong_id() {
+	env_logger::init();
+	let server =
+		WebSocketTestServer::with_hardcoded_response("127.0.0.1:0".parse().unwrap(), internal_error(Id::Num(99_u64)))
+			.await;
+	let uri = to_ws_uri_string(server.local_addr());
+	let client = WsClient::new(&uri, WsConfig::default()).await.unwrap();
+	let err: Result<jsonrpc::JsonValue, Error> = client.request("say_hello", jsonrpc::Params::None).await;
+	assert!(matches!(err, Err(Error::TransportError(e)) if e.to_string().contains("background task closed")));
```

Comment:
This is error is quite bad, we just terminate the background thread and the [frontend](https://github.com/paritytech/jsonrpsee/blob/v2/src/client/ws/client.rs#L182-#L185) will receive https://docs.rs/futures/0.3.8/futures/channel/oneshot/struct.Canceled.html.

We could introduce another `oneshot channel` to send error in or find the pending request/subscription and send an error on that channel.